### DISCUSSION TITLE
Possible issue with 2.6.3rc1 - '{run_id}' is invalid

### DISCUSSION THREAD

**Discussion by lewijw** [apache/airflow#32468] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Our integration test passes with 2.6.2, but fails with 2.6.3rc1 with this message.  This might be our issue, but I thought I would let you know.   Our integration test uses the following id in the URL:

integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad

In the same POST, this Name is passed into the body:

An instance of integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad at 2023-07-07 20:06:57.64
2492 -0400 EDT

The new version seems to taking this Name over the dag id that is passed into the URL.  Here is the message:

    f"The provided run ID '{run_id}' is invalid. It does not match either "
airflow.exceptions.AirflowException: The provided run ID 'An instance of integration_test.
sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad at 2023-07-07 20:06:57.64
2492 -0400 EDT' is invalid. It does not match either the configured pattern: '^[A-Za-z0-9_
.~:+-]+$' or the built-in pattern: '^(?:manual|scheduled|dataset_triggered)__(?:\d{4}-\d{2
}-\d{2}T\d{2}:\d{2}:\d{2}\+00:00)$'
10.254.247.184 - Admin [08/Jul/2023:00:19:03 +0000] "POST /api/v1/dags/integration_test.sa
s_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad/dagRuns HTTP/1.1" 500 4535

### What you think should happen instead

It seems the Name in the body should not preempt the id in the URL. 

### How to reproduce

Do a POST /api/v1/dags/{id}/dagRuns with a space in the Name in the body.

### Operating System

RHEL8

### Versions of Apache Airflow Providers

None.

### Deployment

Other

### Deployment details

k8s with kustomize

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by eladkal** [comment]
cc @ephraimbuddy

**Comment by ephraimbuddy** [comment]
Updating the config `[scheduler]allowed_run_id_pattern` will allow you to use any kind of `run_id`. It's not a breaking change

**Comment by lewijw** [comment]
That is interesting.  I did not know that was possible.  However, I think the bigger issue is that the Name in the body is being validated with the same validation intended for the id.  Is that intentional?

**Comment by potiuk** [comment]
> That is interesting. I did not know that was possible. However, I think the bigger issue is that the Name in the body is being validated with the same validation intended for the id. Is that intentional?

What do you mean by "Name in the body"?  As far as I can see the validation is only failing for `run_id` and there is no "Name" in the body? Or am I missing something?

**Comment by lewijw** [comment]
I am sorry I am not being clear.  The error message indicates that this string is the problem:

"An instance of integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad at 2023-07-07 20:06:57.642492 -0400 EDT"

That string is being passed in as the Name in the body of the POST (that may be an error on our part).  The dag_id that is passed in the URL is a smaller string:

integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad

Since the error message is mentioning the larger string that starts with "An instance", it looks like the Name is being subjected to the dag_id validation.

**Comment by potiuk** [comment]
> "An instance of integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad at 2023-07-07 20:06:57.642492 -0400 EDT"
> That string is being passed in as the Name in the body of the POST (that may be an error on our part). The dag_id that is passed in the URL is a smaller string:

There is no "name" in the body. this string is passed as run_id

**Comment by potiuk** [comment]
Simply speaking - your `run_id` generated via integration test of yours "looks like" name, but this is an ID of the run. 

A bit more explanation why it has changed:

It used to be "free form", but it causes a lot of problem (including security issues) that's why we are limiting it to be follow a "real" id format (and you can define how your "id" should look like by providing your own regular exception in configuration).

**Comment by potiuk** [comment]
Converting this one to discussion in case more discussion is needed.

**Comment by potiuk** [comment]
BTW. Thanks for raising the issue @lewijw -> it's really cool that you are running integration tests for RC versions of Airflow. This one is not  valid, really but it's fantastic to see that people are testing our RCs.

**Comment by lewijw** [comment]
You are welcome.  Just trying to help.  I did note at the beginning that I think this is an issue on our part.  You are right that there is no name in the body according to the REST API, and we are sending a Name in error.  It is still very interesting to me that somehow that name is being pulled out and validated as an id.  I am happy keeping this as a discussion.  Thanks for the replies.  I will send the exact body we are sending later - probably Tuesday.

**Reply by potiuk** [reply]
Oh yeah. That's interesting - but I presume, you are also sending it as "run_id". But yeah - please send us the POST, maybe there is a path where some extra parmeters are being interpreted as other parameters, though I'd find it rather strange.

**Comment by lewijw** [comment]
The error message shows this URL:

POST /api/v1/dags/integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad/dagRuns

I believe that is a valid run_id.

**Reply by potiuk** [reply]
Apparently you override it in your POST

**Reply by potiuk** [reply]
That's why it would be good to see it.

**Comment by lewijw** [comment]
I decided to fire up the cluster.  You are right.  Here is the body:

{"dag_run_id":"An instance of integration_test.sas_po_hello_world_poc_flow-5cf6ded1-3949-44da-9f6b-9e59839d99ad at 2023-07-09 17:05:04.542429 -0400 EDT"}

I could have given my left arm that we were sending Name, but I was mistaken.  Actually, the value is coming from a Name column in our database.

So, 2.6.3rc1 is correctly highlighting a bug in our code that 2.6.2 does not.  Thanks again for your replies.

**Reply by potiuk** [reply]
Glad it's been sorted. And keep up testing RCs of Airflow - this is I think super-beneficial for everyone - both we can see errors we missed and you can improve your test harness :)

**Comment by lewijw** [comment]
I plan to do so.  Thanks.

**Comment by lewijw** [comment]
My apologies, I was getting confused thinking the dag_run_id in the body was supposed to be the same as the dag_id, but I now see that we are intentionally setting the dag_run_id to be different.  I am new to Airflow.

I believe this is the related change:  https://github.com/apache/airflow/pull/32293

Unfortunately, I cannot see in that PR or in the release notes as to what problems occurred when the run id was allowed to be free form.  Can you please point me to some information?  I understand from @ephraimbuddy that we can make a change to the configuration to make it free form again, but should we?  I'd like more information as to the vulnerabilities we may expose ourselves to.  Thanks.

**Reply by potiuk** [reply]
This has intentionally been left out from the PR and documentaion because that was a CVE advisory (Security issue). It has been announced today at users@airflow.apache.org (if you are not subscribed there, you should be - see our https://airflow.apache.org/community/ page to find out how). This is how you get the information - by actively watching there.

You can see more details by following the PR listed in the announcement.

It will be announced on the ASF announce@ mailing list as well.

The announcement on users@ is here: https://lists.apache.org/thread/rxddqs76r6rkxsg1n24d029zys67qwwo

**Comment by lewijw** [comment]
Thanks.  Is the CVE number correct, or does it just take a while to get it public?  This page shows it as Not Found:
https://nvd.nist.gov/vuln/detail/CVE-2023-22887

**Reply by potiuk** [reply]
Yes it takes a while
